### ISSUE TITLE
Proposal: Add functionality to return List of error result with Result class

### ISSUE THREAD

**Issue by nagtilaklaxman** [vkhorikov/CSharpFunctionalExtensions#428]

**Comment by ProphetLamb** [comment]
https://github.com/vkhorikov/CSharpFunctionalExtensions/pull/378#issuecomment-1176496173
Gave it some thought.
A new `IResult` compatible, convertible monad could make sense. Where `ResultCollection` is a category of higher order of `IResult`.

**Comment by jeffward01** [comment]
> [#378 (comment)](https://github.com/vkhorikov/CSharpFunctionalExtensions/pull/378#issuecomment-1176496173) Gave it some thought. A new `IResult` compatible, convertible monad could make sense. Where `ResultCollection` is a category of higher order of `IResult`.

It would also be cool to add `IReason` of why it failed, and also `IError`, such as what this library has: [FluentResults](https://github.com/altmann/FluentResults)

**Comment by vkhorikov** [comment]
Yes, he actually was one of the early contributors to this library.

Regarding `IReason` and `IError` -- I'd like to hear more about use cases for these classes.

**Comment by jeffward01** [comment]
> Yes, he actually was one of the early contributors to this library.

Small world, its very cool to see how the ecosystem grows.  Thank you for this @vkhorikov 

Here is an interesting chat they are having over at `FluentResults` regarding 'railway programming': https://github.com/altmann/FluentResults/issues/108 

> Regarding `IReason` and `IError` -- I'd like to hear more about use cases for these classes.

I agree, I will do some research on this and get back to everyone -- if anyone is faster than me, please do the same

**Comment by vkhorikov** [comment]
Linking a related question: https://github.com/vkhorikov/CSharpFunctionalExtensions/discussions/427
